### ISSUE TITLE
[Light client] Disconnect all its peers subsequently to an eth_call

### ISSUE THREAD

**Issue by Tbaut** [paritytech/parity-ethereum#9551]
- **Parity Ethereum version**:v2.2.0-unstable-6b39131-20180913
- **Operating system**: Linux
- **Installation**:  built from source
- **Fully synchronized**: yes
- **Network**:  kovan
- **Restarted**: yes

How to reproduce :
Launching parity with:
`parity --light --chain kovan -l on_demand,rpc`
and calling the bellow RPC a coupe times (it hapened after 4-5 times), it ends up in a state where all peers have been disconnected.
edit: I sometimes can't reproduce it even after ~20 calls

```
$ curl --data '{"method":"eth_call","params":[{"to":"0x4a6e6c3868a279e1d9047b42c3fb356ff4680003","data":"0x70a08231000000000000000000000000004702bdcc3c7dbffd943136107e70b827028600"},"latest"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
{"jsonrpc":"2.0","result":"0x0000000000000000000000000000000000000000019d944653184202a34c0000","id":1}
```
After getting the above result (all good), I get:
```
$ curl --data '{"method":"eth_call","params":[{"to":"0x4a6e6c3868a279e1d9047b42c3fb356ff4680003","data":"0x70a08231000000000000000000000000004702bdcc3c7dbffd943136107e70b827028600"},"latest"],"id":1,"jsonrpc":"2.0"}' -H "Content-Type: application/json" -X POST localhost:8545
{"jsonrpc":"2.0","error":{"code":-32065,"message":"Timeout for On-demand query; 1 query attempts remain for query #0"},"id":1}
```

This makes the light client disconnect to all its peers (from originally 5 to 0), [here are the full logs with `rpc` and `on_demand` traces](https://gist.github.com/Tbaut/8b8ea8e3c7ae733aab6f499836eb6849)

**Comment by Tbaut** [comment]
Note that I can **not** reproduce it all the time :/

**Comment by niklasad1** [comment]
I haven't been able to reproduce this yet

**Comment by Tbaut** [comment]
One thing that can help reproduce it very quickly:
- launch a light client as stated above : `parity --light --chain kovan -l on_demand,rpc`
- launch/install [fether](https://github.com/paritytech/fether/releases)
- add some accounts, add some tokens (anything like Gavcoin) and navigate around, no need to have eth or token.

Fether sends a lot of requests and I generally I am able to freeze it in 30sec max.
Ask me anything if you can't reproduce.

**Comment by niklasad1** [comment]
@Tbaut 

Ok, I have no real idea why it is dropping all its peer maybe related to #9526 

However the freezing is probably because: 

technically every execution request on a non-existent account will end up in a `BadProof`.
If the the account is not found the `provider (full node, runs the execution)`  will create a default account with zero balance! 
Because the provider can't distinguish between different executions (read-only and things that actually changes the state) such as ```eth_call``` vs ```eth_sendTransaction```!
The execution will always fail because of zero balance on the account.
and it will always send `BadProof (which doesn't say anything about the internal execution error)` and the light client waits for a ```"valid response"```until it times out!

So, should we exponse a new request type `ReadOnlyExecution`, drop a execution when we get a `BadProof` or something else?

**Comment by Tbaut** [comment]
^ cc @tomaka or @cheme, @ordian what do you think ?

**Comment by cheme** [comment]
About the `BadProof` I am thinking of a possible race : 
- before ondemand execution send its query (ondemand execution is always readonly), `proved_execution` store block_id::latest in envinfo, but also fetch the headers of the latest block B1, those are stored in `ExecuteParams` and in `TransactionProof` for later use.
- the ondemand execution query is sent to distant peer, in-between there is a new block 
- distant peer execute on the new block B2 (blockid being 'latest')
- when we get its values and try to check proof we use previously initialised `TransactionProof` corresponding to B1 state -> BadProof

This seems like a possibility to me (I may have miss something in my code reading). The log got a block import before failing which means the occurrence of bad proof were close to a new block, it would be interesting to confirm it with other logs.

**Comment by cheme** [comment]
I got some positive results, running https://github.com/cheme/parity-ethereum/commit/39ab72eb62f7e054966298699521e0d335aabd0f . So there is probably also the race problem from my previous comment to fix. Not sure about the way to fix it properly (in my branch hacky fix, we may be querying a block before the latest).

**Comment by niklasad1** [comment]
@cheme 

I'm fairly confident that #9591 solves this issue and it actually matches the behavior that we have in the full-node implementation, see https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/client/client.rs#L1501 and https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/client/client.rs#L1244

**Comment by Tbaut** [comment]
I'm not sure it's the reason @ordian.
Here is a `pip` trace going from 6 peers to 0. And it looks like I still get credits

```bash
2018-09-24 11:23:44  IO Worker #1 INFO import     6/50 peers   2 MiB cache 0 bytes queue  RPC:  1 conn,    7 req/s,  152 µs
2018-09-24 11:23:45  IO Worker #3 TRACE pip  Sending status to peer 56
2018-09-24 11:23:45  IO Worker #1 TRACE pip  Peer 56 disconnecting
2018-09-24 11:23:48  IO Worker #2 TRACE pip  Incoming packet 1 from peer 34
2018-09-24 11:23:48  IO Worker #2 TRACE pip  requesting from peer 34. Cost: 1330496; Available: 30000000000
2018-09-24 11:23:48  IO Worker #0 TRACE pip  Incoming packet 1 from peer 47
2018-09-24 11:23:48  IO Worker #3 TRACE pip  Incoming packet 1 from peer 24
2018-09-24 11:23:48  IO Worker #3 TRACE pip  Incoming packet 1 from peer 45
2018-09-24 11:23:48  IO Worker #1 TRACE pip  Incoming packet 1 from peer 19
2018-09-24 11:23:48  IO Worker #2 TRACE pip  Sending status to peer 14
2018-09-24 11:23:48  IO Worker #0 TRACE pip  Peer 14 disconnecting
2018-09-24 11:23:48  IO Worker #0 TRACE pip  Incoming packet 1 from peer 36
2018-09-24 11:23:48  IO Worker #2 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:48  IO Worker #2 TRACE pip  pre-verifying response for Request #477 from peer 34
2018-09-24 11:23:48  IO Worker #2 TRACE pip  requesting from peer 19. Cost: 26884; Available: 30000000000
2018-09-24 11:23:49  IO Worker #0 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:49  IO Worker #0 TRACE pip  pre-verifying response for Request #478 from peer 19
2018-09-24 11:23:49  Verifier #0 INFO import  Imported #8838535 0xd5f1…e978 (0.16 Mgas)
2018-09-24 11:23:49  IO Worker #1 TRACE pip  Sending status to peer 7
2018-09-24 11:23:49  IO Worker #1 TRACE pip  Peer 7 disconnecting
2018-09-24 11:23:51  IO Worker #3 TRACE pip  Sending status to peer 7
2018-09-24 11:23:51  IO Worker #3 TRACE pip  Peer 7 disconnecting
2018-09-24 11:23:51  IO Worker #1 TRACE pip  Sending status to peer 26
2018-09-24 11:23:51  IO Worker #2 TRACE pip  Peer 26 disconnecting
2018-09-24 11:23:52  IO Worker #1 TRACE pip  Sending status to peer 26
2018-09-24 11:23:52  IO Worker #0 TRACE pip  Peer 26 disconnecting
2018-09-24 11:23:52  IO Worker #2 TRACE pip  Sending status to peer 46
2018-09-24 11:23:52  IO Worker #3 TRACE pip  Peer 46 disconnecting
2018-09-24 11:23:53  IO Worker #2 TRACE pip  Sending status to peer 17
2018-09-24 11:23:53  IO Worker #0 TRACE pip  Peer 17 disconnecting
2018-09-24 11:23:53  IO Worker #0 TRACE pip  Sending status to peer 61
2018-09-24 11:23:53  IO Worker #2 TRACE pip  Peer 61 disconnecting
2018-09-24 11:23:53  IO Worker #1 TRACE pip  Sending status to peer 28
2018-09-24 11:23:53  IO Worker #3 TRACE pip  Peer 28 disconnecting
2018-09-24 11:23:54  IO Worker #2 TRACE pip  Sending status to peer 3
2018-09-24 11:23:54  IO Worker #0 TRACE pip  Peer 3 disconnecting
2018-09-24 11:23:54   TRACE pip  requesting from peer 34. Cost: 606995; Available: 30000000000
2018-09-24 11:23:54   TRACE pip  requesting from peer 34. Cost: 606995; Available: 29999393005
2018-09-24 11:23:54   TRACE pip  requesting from peer 19. Cost: 502013; Available: 30000000000
2018-09-24 11:23:54  IO Worker #3 TRACE pip  Sending status to peer 37
2018-09-24 11:23:54  IO Worker #2 TRACE pip  Peer 37 disconnecting
2018-09-24 11:23:54  IO Worker #3 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:54  IO Worker #3 TRACE pip  pre-verifying response for Request #479 from peer 34
2018-09-24 11:23:54   TRACE pip  requesting from peer 34. Cost: 606995; Available: 29998786010
2018-09-24 11:23:54   TRACE pip  requesting from peer 34. Cost: 606995; Available: 29998179015
2018-09-24 11:23:54  IO Worker #2 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:54  IO Worker #2 TRACE pip  pre-verifying response for Request #481 from peer 19
2018-09-24 11:23:55  IO Worker #1 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:55  IO Worker #1 TRACE pip  pre-verifying response for Request #480 from peer 34
2018-09-24 11:23:55  IO Worker #2 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:55  IO Worker #2 TRACE pip  pre-verifying response for Request #482 from peer 34
2018-09-24 11:23:55  IO Worker #1 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:55  IO Worker #1 TRACE pip  pre-verifying response for Request #483 from peer 34
2018-09-24 11:23:55   TRACE pip  requesting from peer 19. Cost: 950000; Available: 29999497987
2018-09-24 11:23:55  IO Worker #2 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:55  IO Worker #2 TRACE pip  pre-verifying response for Request #484 from peer 19
2018-09-24 11:23:55   TRACE pip  requesting from peer 47. Cost: 290483; Available: 30000000000
2018-09-24 11:23:55   TRACE pip  requesting from peer 45. Cost: 239504; Available: 30000000000
2018-09-24 11:23:55   TRACE pip  requesting from peer 34. Cost: 606995; Available: 29997572020
2018-09-24 11:23:55   TRACE pip  requesting from peer 45. Cost: 239504; Available: 29999760496
2018-09-24 11:23:55   TRACE pip  requesting from peer 19. Cost: 502013; Available: 29999050000
2018-09-24 11:23:55   TRACE pip  requesting from peer 24. Cost: 718791; Available: 30000000000
2018-09-24 11:23:55  IO Worker #2 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:55  IO Worker #2 TRACE pip  pre-verifying response for Request #485 from peer 47
2018-09-24 11:23:55  IO Worker #3 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:55  IO Worker #3 TRACE pip  pre-verifying response for Request #487 from peer 34
2018-09-24 11:23:55   TRACE pip  requesting from peer 34. Cost: 850000; Available: 29999393005
2018-09-24 11:23:55  IO Worker #0 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:55  IO Worker #0 TRACE pip  pre-verifying response for Request #490 from peer 24
2018-09-24 11:23:55   TRACE pip  requesting from peer 45. Cost: 1150000; Available: 29999520992
2018-09-24 11:23:55  IO Worker #3 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:55  IO Worker #3 TRACE pip  pre-verifying response for Request #486 from peer 45
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:56  IO Worker #0 TRACE pip  pre-verifying response for Request #489 from peer 19
2018-09-24 11:23:56  IO Worker #2 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:56  IO Worker #2 TRACE pip  pre-verifying response for Request #488 from peer 45
2018-09-24 11:23:56   TRACE pip  requesting from peer 47. Cost: 290483; Available: 29999709517
2018-09-24 11:23:56   TRACE pip  requesting from peer 24. Cost: 718791; Available: 29999281209
2018-09-24 11:23:56   TRACE pip  requesting from peer 24. Cost: 718791; Available: 29998562418
2018-09-24 11:23:56   TRACE pip  requesting from peer 36. Cost: 528147; Available: 30000000000
2018-09-24 11:23:56   TRACE pip  requesting from peer 36. Cost: 528147; Available: 29999471853
2018-09-24 11:23:56  IO Worker #3 TRACE pip  Incoming packet 1 from peer 47
2018-09-24 11:23:56  IO Worker #3 TRACE pip  requesting from peer 47. Cost: 4100224; Available: 29999419034
2018-09-24 11:23:56   TRACE pip  requesting from peer 24. Cost: 718791; Available: 29997843627
2018-09-24 11:23:56  IO Worker #1 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:56  IO Worker #1 TRACE pip  pre-verifying response for Request #491 from peer 34
2018-09-24 11:23:56  IO Worker #1 TRACE pip  Incoming packet 1 from peer 36
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:56  IO Worker #0 TRACE pip  pre-verifying response for Request #493 from peer 47
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:56  IO Worker #0 TRACE pip  pre-verifying response for Request #492 from peer 45
2018-09-24 11:23:56  IO Worker #1 TRACE pip  Incoming packet 1 from peer 24
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 3 from peer 36
2018-09-24 11:23:56  IO Worker #0 TRACE pip  pre-verifying response for Request #496 from peer 36
2018-09-24 11:23:56  IO Worker #3 TRACE pip  Incoming packet 1 from peer 34
2018-09-24 11:23:56  IO Worker #1 TRACE pip  Incoming packet 3 from peer 36
2018-09-24 11:23:56  IO Worker #1 TRACE pip  pre-verifying response for Request #497 from peer 36
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 1 from peer 19
2018-09-24 11:23:56  IO Worker #2 TRACE pip  Incoming packet 1 from peer 45
2018-09-24 11:23:56  IO Worker #3 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:56  IO Worker #3 TRACE pip  pre-verifying response for Request #494 from peer 24
2018-09-24 11:23:56  IO Worker #0 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:56  IO Worker #0 TRACE pip  pre-verifying response for Request #495 from peer 24
2018-09-24 11:23:56   TRACE pip  requesting from peer 19. Cost: 950000; Available: 29998547987
2018-09-24 11:23:56  IO Worker #3 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:56  IO Worker #3 TRACE pip  pre-verifying response for Request #498 from peer 47
2018-09-24 11:23:56  IO Worker #3 TRACE pip  requesting from peer 24. Cost: 35934; Available: 29997124836
2018-09-24 11:23:56   TRACE pip  requesting from peer 45. Cost: 239504; Available: 29998370992
2018-09-24 11:23:56   TRACE pip  requesting from peer 24. Cost: 718791; Available: 29997088902
2018-09-24 11:23:56   TRACE pip  requesting from peer 45. Cost: 239504; Available: 29998131488
2018-09-24 11:23:56   TRACE pip  requesting from peer 19. Cost: 502013; Available: 29997597987
2018-09-24 11:23:56   TRACE pip  requesting from peer 36. Cost: 528147; Available: 29999471853
2018-09-24 11:23:56   TRACE pip  requesting from peer 47. Cost: 290483; Available: 29995899776
2018-09-24 11:23:56  IO Worker #1 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:56  IO Worker #1 TRACE pip  pre-verifying response for Request #499 from peer 24
2018-09-24 11:23:56   TRACE pip  requesting from peer 34. Cost: 850000; Available: 29998543005
2018-09-24 11:23:56  IO Worker #2 TRACE pip  Incoming packet 3 from peer 36
2018-09-24 11:23:56  IO Worker #2 TRACE pip  pre-verifying response for Request #506 from peer 36
2018-09-24 11:23:56  IO Worker #3 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:56  IO Worker #3 TRACE pip  pre-verifying response for Request #507 from peer 47
2018-09-24 11:23:57   TRACE pip  requesting from peer 47. Cost: 1700000; Available: 29995609293
2018-09-24 11:23:57  IO Worker #1 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:57  IO Worker #1 TRACE pip  pre-verifying response for Request #501 from peer 24
2018-09-24 11:23:57  Verifier #2 INFO import  Imported #8838537 0xcd9e…3507 (0.11 Mgas)
2018-09-24 11:23:57  IO Worker #2 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:57  IO Worker #2 TRACE pip  pre-verifying response for Request #500 from peer 19
2018-09-24 11:23:57  IO Worker #3 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:57  IO Worker #3 TRACE pip  pre-verifying response for Request #502 from peer 45
2018-09-24 11:23:57  IO Worker #0 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:57  IO Worker #0 TRACE pip  pre-verifying response for Request #509 from peer 47
2018-09-24 11:23:57  IO Worker #3 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:57  IO Worker #3 TRACE pip  pre-verifying response for Request #503 from peer 24
2018-09-24 11:23:57  IO Worker #1 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:57  IO Worker #1 TRACE pip  pre-verifying response for Request #508 from peer 34
2018-09-24 11:23:57  IO Worker #1 TRACE pip  Incoming packet 4 from peer 34
2018-09-24 11:23:57  IO Worker #1 TRACE pip  Received an update to request credit params from peer 34
2018-09-24 11:23:57  IO Worker #2 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:57  IO Worker #2 TRACE pip  pre-verifying response for Request #504 from peer 45
2018-09-24 11:23:57   TRACE pip  requesting from peer 36. Cost: 1350000; Available: 29998943706
2018-09-24 11:23:57  IO Worker #2 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:57  IO Worker #2 TRACE pip  pre-verifying response for Request #505 from peer 19
2018-09-24 11:23:57  IO Worker #0 TRACE pip  Incoming packet 3 from peer 36
2018-09-24 11:23:57  IO Worker #0 TRACE pip  pre-verifying response for Request #510 from peer 36
2018-09-24 11:23:57  IO Worker #0 TRACE pip  Initiating disable of peer 36
2018-09-24 11:23:57  IO Worker #0 TRACE pip  requesting from peer 47. Cost: 1700000; Available: 29993909293
2018-09-24 11:23:57  IO Worker #1 TRACE pip  Peer 36 disconnecting
2018-09-24 11:23:57  IO Worker #2 TRACE pip  Incoming packet 3 from peer 47
2018-09-24 11:23:57  IO Worker #2 TRACE pip  pre-verifying response for Request #511 from peer 47
2018-09-24 11:23:57  IO Worker #2 TRACE pip  Initiating disable of peer 47
2018-09-24 11:23:57  IO Worker #2 TRACE pip  requesting from peer 24. Cost: 3400000; Available: 29999245275
2018-09-24 11:23:57  IO Worker #0 TRACE pip  Peer 47 disconnecting
2018-09-24 11:23:57  IO Worker #0 TRACE pip  Incoming packet 3 from peer 24
2018-09-24 11:23:57  IO Worker #0 TRACE pip  pre-verifying response for Request #512 from peer 24
2018-09-24 11:23:58  IO Worker #0 TRACE pip  Initiating disable of peer 24
2018-09-24 11:23:58  IO Worker #0 TRACE pip  requesting from peer 24. Cost: 3400000; Available: 29996600000
2018-09-24 11:23:58  IO Worker #3 TRACE pip  Peer 24 disconnecting
2018-09-24 11:23:58  IO Worker #3 TRACE pip  requesting from peer 45. Cost: 1150000; Available: 29999520992
2018-09-24 11:23:58  IO Worker #3 TRACE pip  Incoming packet 3 from peer 45
2018-09-24 11:23:58  IO Worker #3 TRACE pip  pre-verifying response for Request #514 from peer 45
2018-09-24 11:23:58  IO Worker #3 TRACE pip  Initiating disable of peer 45
2018-09-24 11:23:58  IO Worker #3 TRACE pip  requesting from peer 34. Cost: 800000; Available: 30000000000
2018-09-24 11:23:58  IO Worker #2 TRACE pip  Peer 45 disconnecting
2018-09-24 11:23:58  IO Worker #2 TRACE pip  Incoming packet 3 from peer 34
2018-09-24 11:23:58  IO Worker #2 TRACE pip  pre-verifying response for Request #515 from peer 34
2018-09-24 11:23:58  IO Worker #2 TRACE pip  Initiating disable of peer 34
2018-09-24 11:23:58  IO Worker #2 TRACE pip  requesting from peer 34. Cost: 800000; Available: 29999200000
2018-09-24 11:23:58  IO Worker #3 TRACE pip  Peer 34 disconnecting
2018-09-24 11:23:58  IO Worker #3 TRACE pip  requesting from peer 19. Cost: 950000; Available: 30000000000
2018-09-24 11:23:59  IO Worker #3 TRACE pip  Incoming packet 3 from peer 19
2018-09-24 11:23:59  IO Worker #3 TRACE pip  pre-verifying response for Request #517 from peer 19
2018-09-24 11:23:59  IO Worker #3 TRACE pip  Initiating disable of peer 19
2018-09-24 11:23:59  IO Worker #3 TRACE pip  requesting from peer 19. Cost: 950000; Available: 29999050000
2018-09-24 11:23:59  IO Worker #2 TRACE pip  Peer 19 disconnecting
2018-09-24 11:24:01  IO Worker #2 TRACE pip  Sending status to peer 41
2018-09-24 11:24:01  IO Worker #0 TRACE pip  Peer 41 disconnecting
2018-09-24 11:24:19  IO Worker #0 INFO import     0/50 peers   2 MiB cache 0 bytes queue  RPC:  1 conn,    8 req/s,  142 µs
```
